### PR TITLE
Feat(app-platform) 3rd party initiated installation

### PR DESCRIPTION
**Pull request by scefali**
This PR implements a new installation flow for Sentry Apps that can start on the website of a third party. The third-party website can expose a link of the following format: `https://sentry.io/sentry-apps/<sentryAppSlug>/external-install/`.

If the user is a member of a single organization, we automatically assume they want to install the integration on that org and don't show an org select:

Screen Shot 2019-08-21 at 9 32 05 AM

If the user is in multiple orgs, we show a dropdown to pick the org before we render more information about the app:

Screen Shot 2019-08-20 at 4 27 09 PM

Once the user selects an org, we load the full information about that org and show the full integration install view:

Screen Shot 2019-08-21 at 9 40 03 AM

Note there are a number of error scenarios which are covered:

- User does not have access to Sentry App or Sentry App doesn't exist (user sees the default 404 page)
- Sentry App is internal
- Sentry App already installed
- Sentry App is an unpublished integration for a different org than is selected
- User does not have permission to install app

See more details [here](https://www.notion.so/sentry/Third-Party-Initiated-Installation-a2acd2601c0c4445a51b20aacdc15153)

### CODE REVIEW

**Review by MeredithAnya — CHANGES_REQUESTED** [review]
So when testing this out, it seems like when you change the dropdown to an organization that already has the app installed, it first shows 'Accept & Install' and then updates to show the error/gray out the button. However there was enough time for someone to actually click the button and in doing so I was able to install the same integration twice on an org.

I think we need to handle this in the mediator OR add the unique constraint on the organization_id and the sentry_app_id for the SentryAppInstallation model

File: src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.jsx [hunk 1]
```diff
@@ -42,6 +42,11 @@ export default class SentryAppDetailsModal extends AsyncComponent {
     onInstall: PropTypes.func.isRequired,
     isInstalled: PropTypes.bool.isRequired,
     closeModal: PropTypes.func.isRequired,
+    closeOnInstall: PropTypes.bool.isRequired,
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.jsx]: @MeredithAnya We don't need it if `onInstall` is redirecting us to another page. Is the issue that the prop type is required?

File: src/sentry/static/sentry/app/views/sentryAppExternalInstallation.jsx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  get disableInstall() {
+    const {reloading, isInstalled} = this.state;
+    return isInstalled || reloading || this.isSentryAppUnavailableForOrg;
+  }
+
+  hasAccess = org => org.access.includes('org:integrations');
+
+  onClose = () => {
+    //if we came from somewhere, go back there. Otherwise, back to the root
+    const newUrl = document.referrer || '/';
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/static/sentry/app/views/sentryAppExternalInstallation.jsx]: @MeredithAnya Good suggestion. Let me make that change.

File: src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.jsx]: @MeredithAnya My bad, I think I gave you an incorrect explanation. The issue without the change is that even if the redirectUrl is given, `onClose` is still called. Meaning you don't actually redirect to the right place during installation. Adding the field `closeOnInstall` and not calling `closeModal` lets `onInstall` handle the redirect after installation. This wasn't an issue before the integrations page just has a popup and it didn't matter that the popup was closed after uninstalling but it does matter if the close behavior changes the current page.

**Review by MeredithAnya — APPROVED** [review]

**Pull request merged**